Dynamically build help text for CLI

ixs opened the pull request:
Instead of hardcoding the help text in the do_help() function,
build the help text dynamically from the docstring of the do_*()
functions.

Makes it easier to extend the CLI.

ixs committed:
Dynamically build help text for CLI

Instead of hardcoding the help text in the do_help() function,
build the help text dynamically from the docstring of the do_*()
functions.

Makes it easier to extend the CLI.

knorrie commented:
Neat! I like this. It puts all the stuff for a command in one place.

1) I also like to be able to have help sorted in a particular order. Does this mean that we'll have to reorder the do_ functions in the code? If the order of the resulting help text is predictable(!) I don't mind doing this one-time shuffle. *)
2) What happens when I have a multi-line doc string? (It will include everything). What about only using the first line? (or, the second line if the first line is only """\s*\n and the actual first line of text is on the second line)?
3) Help text is still optional, so it will not help/force to remind you to add help text when adding a new cli command. I think this is ok, since this might be on purpose (obscure debugging command) or for stuff like do_EOF we don't want extra crap in the help text.

*) When I write such a help section, I try to imagine I'm a new user reading the help lines. So, I want to see most relevant things first, and have things in a logical order to visually process, or, to try out first! That means, on and off, and lock and unlock right after each other, login at the very beginning (or not, because the tool forces you to login first always), and logout and exit command at the end, and obscure commands that you mostly don't need somewhere way at the bottom. If this needs a reshuffle of a lot of cli commands code, then do that first in a git commit, and then do moving the help text into the functions and the helper (the current commit) afterwards.

I think that when reviewing it again, this would be my preferred order, following that pattern of thought:

status <- wtf did I just connect/login to, what can I do here?
listipdus <- ah ok, so there's multiple pdus I can see outlets of
on <- let's enable one
off <- blah
cycle
lock <- etc...
unlock
save <- oh, this is actually really useful to do before exiting, otherwise the PDU will happily shut down the port after a pdu software reboot (been there, did not save, had outage)
exit <- kthxbye
help <- this thing
sort <- I think we should remove this, noone wants unsorted stuff
login <- happens when starting, you probably don't need this
logout <- you also probably do not want to manually do this

knorrie commented on src/pmCommand.py:
```diff
@@ -122,23 +135,14 @@ def emptyline(self):
+
+        for name, func_obj in sorted(inspect.getmembers(self)):
+            if name.startswith('do_') and func_obj.__doc__:
+                print("  %-20s%s" % (name[3:], func_obj.__doc__))
```
I should add some file with hints about [coding style](https://help.github.com/en/github/building-a-strong-community/setting-guidelines-for-repository-contributors#adding-a-contributing-file).

It will include:
* Preferred string formatting method in here is .format(

ixs committed:
Rework help function

inspect does not return modules in the order they are written down in the code.
Use a helper variable pref_order to keep the order we preferr.
Fetch the first line of the docstring from the method and use that.
This allows for additional docstring text without clobbering the help()
function.

ixs commented:
There's a second commit now, doing this. It's a bit messier now, because we basically need to iterate twice...

First we just get the docstrings from the functions in the order we defined earlier.
Then we do the remaining methods...

knorrie commented:
Hi, thanks for your patience. Apologies for taking a while to finally return back here.

If you don't like running through the function list twice, in two different ways, and almost duplicate code to split and format in different ways then you can start with retrieving all the help text and then only printing is left. For performance, it's of course not needed, but it's a little less complex for whoever wants to read this code later indeed.

Spoiler alert, a possible way: https://paste.debian.net/hidden/d31849fc/ (Sorry, this is fun)

knorrie commented:
Are the unrelated formatting changes in do_login and do_logout on purpose? Why?
